docs(metrics): docker container metrics (#1017)
SigNoz/signoz.io: data/docs/metrics-management/docker-container-metrics.mdx

---
date: 2024-12-08
title: Docker container metrics
id: docker-container-metrics
---

This document provides an explains how to monitor Docker container Metrics using OTel Collector and SigNoz.

## Pre-requisites

> - [OTel Collector](/docs/tutorial/opentelemetry-binary-usage-in-virtual-machine/).

Reviewer: This should be the first step of the document instead of a pre-requisite.

You can remove it from here and create a section just after Pre-requisites

## Setup OTel Collector

- Docker API version 1.22+
- Linux (`docker_stats` receiver does not work on Windows or Mac).

> - Either a [SigNoz Cloud account](/docs/cloud/) or a [Self-Hosted SigNoz instance](/docs/install/docker/).

Reviewer: ```suggestion
```

### Configuring the processor

To enrich the collected metrics with additional metadata like hostname and OS, you will need to configure Docker Resource Detection Processor by adding the following configuration to the OTel Collector config file:

```yaml {2-5}
processors:
  resourcedetection/docker:
    detectors: [env, docker]
    timeout: 2s
    override: false
```

### Configuring the pipelines

Once the receiver and processor is configured, make sure to also enable them in the `pipelines` section of the OTel Collector config file:

```yaml {3-4}
service:
  pipelines:
    metrics:
      receivers: [docker_stats]
      processors: [resourcedetection/docker]
```

## Sending Docker metrics to SigNoz

You will also need to configure `exporters` in the OTel Collector config file to send metrics to SigNoz. The configuration will be different based on whether you are using a SigNoz Cloud account or a Self-Hosted SigNoz instance.

### Configuring the exporter

<Tabs>

<TabItem value="cloud" label="SigNoz Cloud" default>

```yaml {2-7}
exporters:
  otlp:
    endpoint: "ingest.{region}.signoz.cloud:443"
    tls:
      insecure: false
    headers:

>       signoz-access-token: "{signoz-token}"

Reviewer: ```suggestion
      signoz-ingestion-key: "{signoz-ingestion-key}"
```

This was recently changed.

```

> You would need to replace `{region}` and `{signoz-token}` in the above file with the region of your choice and token obtained from Signoz Cloud → Settings → Ingestion Settings.

Reviewer: ```suggestion
You would need to replace `{region}` and `{signoz-token}` in the above file with the [region](https://signoz.io/docs/ingestion/signoz-cloud/overview/) of your SigNoz Cloud environment and [Ingestion Key ](https://signoz.io/docs/ingestion/signoz-cloud/keys/)
```

<figure data-zoomable align='center'>
    <img src="/img/docs/metrics/docker/network_bytes.webp" width="80%" alt="Network Bytes"/>
    <figcaption><i>Network Bytes Panels</i></figcaption>
</figure>

Additionally, there's a panel that shows the number of packets dropped by the containers, which uses `container_network_io_usage_rx_dropped` and `container_network_io_usage_tx_dropped` metrics:

<figure data-zoomable align='center'>
    <img src="/img/docs/metrics/docker/network_packets_dropped.webp" width="80%" alt="Network Packets Dropped"/>
    <figcaption><i>Network Packets Dropped Panels</i></figcaption>
</figure>

### Disk Metrics

The dashboard also includes a "Block IO" panel based on `container.blockio.io_serviced_recursive` metric that shows the number of IOs (bio) issued to the block device per second as well as the type of IO operations (read/write):

<figure data-zoomable align='center'>
    <img src="/img/docs/metrics/docker/block_io.webp" width="80%" alt="Block IO"/>
    <figcaption><i>Block IO Panel</i></figcaption>
</figure>

> ## Conclusion

Reviewer: ```suggestion

```
No need for a conclusion section here.
If there are any troubleshooting steps - general or specific that can be helpful for a user, let's add troubleshooting steps. If not, we can let it be.

> Following this guide, you should have a working setup for collecting Docker container metrics using OTel Collector and visualizing them in SigNoz.

Reviewer: ```suggestion

```

> You can find the full list of metrics collected by the `docker_stats` receiver and available for querying in dashboards [in the official receiver documentation](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/dockerstatsreceiver/documentation.md).

Reviewer: Add this as the final sentence after Disk Metrics under the Visualizing Docker Container Metrics section.

> You can find the full list of metrics collected by the `docker_stats` receiver and available for querying in dashboards [in the official receiver documentation](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/dockerstatsreceiver/documentation.md).

Reviewer: Add this as the final sentence after Disk Metrics under the Visualizing Docker Container Metrics section.
